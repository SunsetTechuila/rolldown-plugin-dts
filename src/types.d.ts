declare module 'rolldown' {
  export interface CustomPluginOptions {
    dtsFile?: boolean
  }
}

export {}
