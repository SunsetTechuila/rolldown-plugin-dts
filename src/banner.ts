import MagicString from 'magic-string'
import { RE_DTS } from './filename.ts'
import type { OptionsResolved } from './options.ts'
import type { Plugin } from 'rolldown'

export function createBannerPlugin({
  banner,
  footer,
}: Pick<OptionsResolved, 'banner' | 'footer'>): Plugin {
  return {
    name: 'rolldown-plugin-dts:banner',
    async renderChunk(code: string, chunk) {
      if (!RE_DTS.test(chunk.fileName)) {
        return
      }

      const s = new MagicString(code)
      if (banner) {
        const code = await (typeof banner === 'function'
          ? banner(chunk)
          : banner)
        if (code) s.prepend(`${code}\n`)
      }
      if (footer) {
        const code = await (typeof footer === 'function'
          ? footer(chunk)
          : footer)
        if (code) s.append(`\n${code}`)
      }

      return {
        code: s.toString(),
        get map() {
          return s.generateMap({
            source: chunk.fileName,
            includeContent: true,
            hires: 'boundary',
          })
        },
      }
    },
  }
}
