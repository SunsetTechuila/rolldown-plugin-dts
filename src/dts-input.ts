import type { Plugin } from 'rolldown'

export function createDtsInputPlugin(): Plugin {
  return {
    name: 'rolldown-plugin-dts:dts-input',

    options(options) {
      return {
        treeshake:
          options.treeshake !== false
            ? {
                ...(options.treeshake === true ? {} : options.treeshake),
                moduleSideEffects: false,
              }
            : false,
        ...options,
      }
    },

    outputOptions(options) {
      return {
        ...options,
        entryFileNames(chunk) {
          if (chunk.name.endsWith('.d')) {
            return '[name].ts'
          }
          return '[name].d.ts'
        },
      }
    },
  }
}
