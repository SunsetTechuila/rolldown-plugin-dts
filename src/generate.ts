import { basename } from 'node:path'
import { isolatedDeclaration as oxcIsolatedDeclaration } from 'oxc-transform'
import {
  filename_dts_to,
  filename_ts_to_dts,
  isRelative,
  RE_DTS,
  RE_NODE_MODULES,
  RE_TS,
} from './utils/filename'
import type { Options } from '.'
import type { Plugin } from 'rolldown'

export function createGeneratePlugin({
  isolatedDeclaration,
  inputAlias = {},
  external,
}: Pick<Options, 'external' | 'isolatedDeclaration' | 'inputAlias'>): Plugin {
  const dtsMap = new Map<string, string>()
  const inputAliasMap = new Map<string, string>(Object.entries(inputAlias))

  let inputOption: Record<string, string> | undefined

  return {
    name: 'rolldown-plugin-dts:generate',

    options({ input }) {
      if (isPlainObject(input)) {
        inputOption = { ...input }
      }
    },

    transform: {
      order: 'pre',
      filter: {
        id: {
          include: [RE_TS],
          exclude: [RE_DTS, RE_NODE_MODULES],
        },
      },
      handler(code, id) {
        const { code: dtsCode, errors } = oxcIsolatedDeclaration(
          id,
          code,
          isolatedDeclaration,
        )
        if (errors.length) {
          return this.error(errors[0])
        }

        const dtsId = filename_ts_to_dts(id)
        dtsMap.set(dtsId, dtsCode)

        const mod = this.getModuleInfo(id)
        if (mod?.isEntry) {
          let fileName = basename(dtsId)

          if (inputAliasMap.has(fileName)) {
            fileName = inputAliasMap.get(fileName)!
          } else if (inputAliasMap.has(dtsId)) {
            fileName = inputAliasMap.get(dtsId)!
          }

          this.emitFile({
            type: 'chunk',
            id: dtsId,
            fileName,
          })
        }
      },
    },

    async resolveId(id, importer, extraOptions) {
      // must be entry
      if (dtsMap.has(id)) {
        return { id, meta: { dtsFile: true } }
      }

      if (importer && this.getModuleInfo(importer)?.meta.dtsFile) {
        if (
          // FIXME external all deps temporarily
          // should introduce custom resolver for resolving types from node_modules
          !isRelative(id) ||
          external?.(id, importer!, extraOptions) === true
        ) {
          return { id, external: true }
        }

        // link to the original module
        const resolution = await this.resolve(
          id,
          filename_dts_to(importer!, 'ts'),
        )
        if (!resolution || resolution.external) return

        const dtsId = filename_ts_to_dts(resolution.id)
        if (dtsMap.has(dtsId)) {
          return { id: dtsId, meta: { dtsFile: true } }
        }

        // pre-load original module if not already loaded
        await this.load(resolution)
        if (dtsMap.has(dtsId)) {
          return { id: dtsId, meta: { dtsFile: true } }
        }
      } else if (extraOptions.isEntry && inputOption) {
        // mapping entry point to dts filename
        const resolution = await this.resolve(id, importer, extraOptions)
        if (!resolution) return

        const dtsId = filename_ts_to_dts(resolution.id)
        if (inputAliasMap.has(dtsId)) return resolution

        for (const [name, entry] of Object.entries(inputOption)) {
          if (entry === id) {
            inputAliasMap.set(dtsId, `${name}.d.ts`)
            break
          }
        }

        return resolution
      }
    },

    load: {
      filter: {
        id: {
          include: [RE_DTS],
          exclude: [RE_NODE_MODULES],
        },
      },
      handler(id) {
        if (dtsMap.has(id)) {
          return {
            code: dtsMap.get(id)!,
            moduleSideEffects: false,
          }
        }
      },
    },
  }
}

function isPlainObject(data: unknown): data is Record<PropertyKey, unknown> {
  if (typeof data !== 'object' || data === null) {
    return false
  }

  const proto = Object.getPrototypeOf(data)
  return proto === null || proto === Object.prototype
}
