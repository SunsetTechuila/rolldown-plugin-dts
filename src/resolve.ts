import { createResolver } from 'dts-resolver'
import {
  filename_js_to_dts,
  filename_ts_to_dts,
  isRelative,
  RE_CSS,
  RE_DTS,
  RE_JS,
  RE_NODE_MODULES,
  RE_TS,
} from './utils/filename'
import type { OptionsResolved } from '.'
import type { Plugin } from 'rolldown'

export const meta = { dtsFile: true } as const

export function createDtsResolvePlugin({
  tsconfig,
  resolve,
  resolvePaths,
}: Pick<OptionsResolved, 'tsconfig' | 'resolve' | 'resolvePaths'>): Plugin {
  const resolver = createResolver({
    tsconfig: tsconfig ? (tsconfig as string) : undefined,
  })

  function resolveDependency(id: string, importer: string) {
    let shouldResolve: boolean
    if (typeof resolve === 'boolean') {
      shouldResolve = resolve
    } else {
      shouldResolve = resolve.some((pattern) =>
        typeof pattern === 'string' ? id === pattern : pattern.test(id),
      )
    }
    if (shouldResolve) {
      const resolution = resolver(id, importer)
      if (resolution) return { id: resolution, meta }
    } else {
      return { id, external: true, meta }
    }
  }

  return {
    name: 'rolldown-plugin-dts:resolve',

    resolveId: {
      order: 'pre',
      async handler(id, importer, options) {
        // only resolve in dts file
        if (
          !importer ||
          (!RE_DTS.test(importer) &&
            !this.getModuleInfo(importer)?.meta.dtsFile)
        ) {
          return
        }

        if (RE_CSS.test(id)) {
          return {
            id,
            external: true,
            moduleSideEffects: false,
          }
        }

        if (RE_NODE_MODULES.test(importer)) {
          const resolution = resolver(id, importer)
          if (resolution) return { id: resolution, meta }
        }

        // resolve dependency [pre]
        if (!resolvePaths && (RE_NODE_MODULES.test(id) || !isRelative(id))) {
          return resolveDependency(id, importer)
        }

        let resolution = await this.resolve(id, importer, options)
        if (!resolution && !id.endsWith('.d')) {
          resolution = await this.resolve(`${id}.d`, importer, options)
        }

        if (resolution?.id) {
          if (RE_CSS.test(resolution.id)) {
            return {
              id,
              external: true,
              moduleSideEffects: false,
            }
          }
          if (resolution.id.startsWith('\0')) {
            return { ...resolution, meta }
          }
        }

        // resolve dependency [post]
        if (
          resolvePaths &&
          (RE_NODE_MODULES.test(resolution?.id || id) ||
            !isRelative(resolution?.id || id))
        ) {
          return resolveDependency(id, importer)
        }

        if (!resolution || resolution.external) {
          return resolution
        }

        if (RE_JS.test(resolution.id)) {
          // resolve js to dts
          resolution = await this.resolve(
            filename_js_to_dts(resolution.id),
            importer,
            options,
          )
          if (!resolution) return
        } else if (RE_TS.test(resolution.id) && !RE_DTS.test(resolution.id)) {
          // FIXME: rolldown bug
          await Promise.any([
            this.load(resolution),
            new Promise((resolve) => setTimeout(resolve, 200)),
          ])

          // redirect ts to dts
          resolution.id = filename_ts_to_dts(resolution.id)
        }

        if (RE_DTS.test(resolution.id)) {
          return { ...resolution, meta }
        }
      },
    },
  }
}
