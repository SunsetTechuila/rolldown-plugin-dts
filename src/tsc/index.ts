import Debug from 'debug'
import { tscEmitBuild } from './emit-build.ts'
import { tscEmitCompiler } from './emit-compiler.ts'
import { ts } from './require-tsc.ts'
import type { TscOptions, TscResult } from './types.ts'

export type { TscModule, TscOptions, TscResult } from './types.ts'

const debug = Debug('rolldown-plugin-dts:tsc')
debug(`loaded typescript: ${ts.version}`)

export function tscEmit(tscOptions: TscOptions): TscResult {
  debug(`running tscEmit ${tscOptions.id}`)

  if (tscOptions.build) {
    return tscEmitBuild(tscOptions)
  } else {
    return tscEmitCompiler(tscOptions)
  }
}
