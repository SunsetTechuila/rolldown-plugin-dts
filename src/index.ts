import { createFakeJsPlugin } from './fake-js'
import { createGeneratePlugin } from './generate'
import type { IsolatedDeclarationsOptions } from 'oxc-transform'
import type { Plugin } from 'rolldown'

export interface Options {
  /**
   * When entries are `.d.ts` files (instead of `.ts` files), this option should be set to `true`.
   *
   * If enabled, the plugin will skip generating a `.d.ts` file for the entry point.
   */
  dtsInput?: boolean

  isolatedDeclaration?: Omit<IsolatedDeclarationsOptions, 'sourcemap'>
  /**
   * dts file name alias `{ [filename]: path }`
   *
   * @example
   * ```ts
   * inputAlias: {
   *   'foo.d.ts': 'foo/index.d.ts',
   * }
   */
  inputAlias?: Record<string, string>

  /** Resolve external types used in dts files from `node_modules` */
  resolve?: boolean | (string | RegExp)[]
}

export function dts(options: Options = {}): Plugin[] {
  const plugins: Plugin[] = []
  if (!options.dtsInput) {
    plugins.push(createGeneratePlugin(options))
  }
  plugins.push(createFakeJsPlugin(options))
  return plugins
}

export { createFakeJsPlugin, createGeneratePlugin }
