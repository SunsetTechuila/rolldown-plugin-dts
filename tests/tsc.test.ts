import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { rolldownBuild } from '@sxzz/test-utils'
import { expect, test } from 'vitest'
import { dts } from '../src'

const dirname = path.dirname(fileURLToPath(import.meta.url))
const root = path.resolve(dirname, 'fixtures/tsc')

test('tsc', async () => {
  const { snapshot } = await rolldownBuild(
    null!,
    [
      dts({
        emitDtsOnly: true,
        compilerOptions: {},
        isolatedDeclaration: false,
      }),
    ],
    {
      input: [path.resolve(root, 'entry1.ts'), path.resolve(root, 'entry2.ts')],
    },
  )
  expect(snapshot).toMatchSnapshot()
})
