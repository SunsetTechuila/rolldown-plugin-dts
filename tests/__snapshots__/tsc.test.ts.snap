// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tsc 1`] = `
"// entry1.d.ts

//#region tests/fixtures/tsc/mod.d.ts
declare const str = "foo";
declare const num = 42;
declare const bool = true;
declare function fn(arg: typeof str): "foo" | 1;

//#endregion
export { bool, fn, num, str };
// entry2.d.ts

//#region tests/fixtures/tsc/entry2.d.ts
interface Unused {
    foo: string;
}

//#endregion
export { Unused };"
`;
